Make GOV.UK Frontend assets paths work with http prefix

lfdebrux opened the pull request:
<!--
## Please fill in the sections below

After you submit your pull request, the technical writing team from the Central Digital and Data Office (CDDO) will discuss and prioritise it at our fortnightly triage meeting. We’ll then let you know if and when we’ll move it forward.
-->

## What’s changed

You no longer need to change the Sass variable `govuk-assets-path` when using a http prefix; the tech docs gem will add the prefix for you.

## Identifying a user need

<!-- Do you have evidence that this meets the needs of users? Let us know about any user research or testing you’ve done. -->

Users who want to deploy their tech docs to GitHub pages need to handle an additional path prefix to all URLs; they may want to use the `http_prefix` setting for this. PR #197 added this http prefix to all assets from the tech docs gem, but required the user to also override `govuk-assets-path` themselves so that assets from GOV.UK Frontend had the correct path. This PR should mean that this additional step is no longer required, making configuration easier for users.

lfdebrux committed:
Update changelog

36degrees reviewed (commented):
If I understand what this is doing correctly, then I think there's a neater way to do this that doesn't need us to monkey patch the `asset_path` function.

GOV.UK Frontend allows us to define which function to use when generating paths for images ([`$govuk-image-url-function`](https://frontend.design-system.service.gov.uk/sass-api-reference/#govuk-image-url-function)) and the font ([`$govuk-font-url-function`](https://frontend.design-system.service.gov.uk/sass-api-reference/#govuk-font-url-function)).

If no custom function is set, the default behaviour is just to prefix the `$filename` with `$govuk-images-path` / `$govuk-fonts-path` (which default to `"#{$govuk-assets-path}images/"` and `"#{$govuk-assets-path}fonts/"` respectively).

I think we can instead pass our own function instead which does the same thing but then calls `asset-path` with the resolved path _including the filename_. So something like this:

```scss
$govuk-assets-path: "/assets/govuk/assets/";

@function tech-docs-image-handler($filename) {
  @return asset-path($govuk-image-url-function + $filename);
}

$govuk-image-url-function: 'tech-docs-image-handler';
```

As ever, happy to talk through it if it's helpful!

lfdebrux committed:
Make GOV.UK Frontend assets paths work with http prefix

Co-authored-by: Oliver Byford <oliver.byford@digital.cabinet-office.gov.uk>

lfdebrux committed:
Add deprecation warning for govuk_assets_path

Setting govuk_assets_path should no longer be necessary, and to keep the
code simple it will no longer have any effect. This commit lets users know
about this change in case it causes them some surprise.

lfdebrux commented:
> If I understand what this is doing correctly, then I think there's a neater way to do this that doesn't need us to monkey patch the `asset_path` function.
> 
> GOV.UK Frontend allows us to define which function to use when generating paths for images ([`$govuk-image-url-function`](https://frontend.design-system.service.gov.uk/sass-api-reference/#govuk-image-url-function)) and the font ([`$govuk-font-url-function`](https://frontend.design-system.service.gov.uk/sass-api-reference/#govuk-font-url-function)).
> 
> If no custom function is set, the default behaviour is just to prefix the `$filename` with `$govuk-images-path` / `$govuk-fonts-path` (which default to `"#{$govuk-assets-path}images/"` and `"#{$govuk-assets-path}fonts/"` respectively).
> 
> I think we can instead pass our own function instead which does the same thing but then calls `asset-path` with the resolved path _including the filename_. So something like this:
> 
> ```css-scss
> $govuk-assets-path: "/assets/govuk/assets/";
> 
> @function tech-docs-image-handler($filename) {
>   @return asset-path($govuk-image-url-function + $filename);
> }
> 
> $govuk-image-url-function: 'tech-docs-image-handler';
> ```
> 
> As ever, happy to talk through it if it's helpful!

Thanks for this @36degrees, this is much nicer!

I've pushed a different version of the code now based on your suggestion.

36degrees commented on lib/assets/stylesheets/_govuk_tech_docs.scss:
```diff
@@ -1,4 +1,15 @@
-$govuk-assets-path: "/assets/govuk/assets/" !default;
+@function tech-docs-image-url-function($filename) {
+  @return image-url($govuk-images-path + $filename)
+}
+
+@function tech-docs-font-url-function($filename) {
+  @return font-url($govuk-fonts-path + $filename)
+}
```
Minor but it looks like these need semicolons for the syntax highlighting to work correctly

```suggestion
@function tech-docs-image-url-function($filename) {
  @return image-url($govuk-images-path + $filename);
}

@function tech-docs-font-url-function($filename) {
  @return font-url($govuk-fonts-path + $filename);
}
```

36degrees replied:
Weird how the syntax highlighting works correctly in the diff 🤷🏻

lfdebrux replied:
Huh, weird. Thanks for catching that, I didn't realise Sass is semicolon-y.

36degrees commented on lib/assets/stylesheets/_govuk_tech_docs.scss:
```diff
@@ -1,4 +1,15 @@
+$govuk-font-url-function: 'tech-docs-font-url-function';
+$govuk-image-url-function: 'tech-docs-image-url-function';
+
+$govuk-assets-path: "/assets/govuk/assets/";
```
I suspect this does work as intended, because presumably `$govuk-fonts-path` isn't evaluated until the function is called, but it might make people's heads hurt less if this variable was set before the functions were defined

(I appreciate that `$govuk-images-path` and `$govuk-fonts-path` haven't been defined at all at this point!)

lfdebrux replied:
Hmm, yes, I don't know, I kind of thought that having it after would help people realise that `$govuk-{images,fonts}-path` are not defined here and that they need to dig deeper to understand it better. Maybe a comment would be more kind though 😂

lfdebrux replied:
I've added some comments, let me know if you think more detail is needed.

36degrees reviewed (commented):
I'm guessing there's no easy way to test this, given the functionality we're interested in is really part of Sprockets?

Otherwise I'm happy with this approach, bar a couple of very minor comments.

lfdebrux commented:
> I'm guessing there's no easy way to test this, given the functionality we're interested in is really part of Sprockets?
> 
> Otherwise I'm happy with this approach, bar a couple of very minor comments.

Ah tests is a good idea, I'll add some.

JC-DSIT commented:
Gonna close this here as it's quite old and needs a review in with the latest changes.  Have added it to issue #489 for tracking
